src/lib/utils/thorchain/index.ts

```diff
@@ -0,0 +1,223 @@
+import { fromBaseUnit, toBaseUnit } from 'lib/math'
+import {
+  type EstimatedFeesQueryKey,
+  queryFn as getEstimatedFeesQueryFn,
+} from 'pages/Lending/hooks/useGetEstimatedFeesQuery'
+import type { IsSweepNeededQueryKey } from 'pages/Lending/hooks/useIsSweepNeededQuery'
+import { queryFn as isSweepNeededQueryFn } from 'pages/Lending/hooks/useIsSweepNeededQuery'
+import { selectPortfolioCryptoBalanceBaseUnitByFilter } from 'state/slices/common-selectors'
+import type { ThorchainSaversWithdrawQuoteResponseSuccess } from 'state/slices/opportunitiesSlice/resolvers/thorchainsavers/types'
+import { fromThorBaseUnit } from 'state/slices/opportunitiesSlice/resolvers/thorchainsavers/utils'
+import { isUtxoChainId } from 'state/slices/portfolioSlice/utils'
+import { selectMarketDataById } from 'state/slices/selectors'
+import { store } from 'state/store'
+
+import type { GetThorchainSaversDepositQuoteQueryKey } from './hooks/useGetThorchainSaversDepositQuoteQuery'
+import { queryFn as getThorchainSaversDepositQuoteQueryFn } from './hooks/useGetThorchainSaversDepositQuoteQuery'
+import {
+  type GetThorchainSaversWithdrawQuoteQueryKey,
+  queryFn as getThorchainSaversWithdrawQuoteQueryFn,
+} from './hooks/useGetThorchainSaversWithdrawQuoteQuery'
+
+// TODO(gomes): this will work for UTXO but is invalid for tokens since they use diff. denoms
+// the current workaround is to not do fee deduction for non-UTXO chains,
+// but for consistency, we should for native EVM assets, and ensure this is a no-op for tokens
+// Note when implementing this, fee checks/deduction will need to either be done for *native* assets only
+// or handle different denoms for tokens/native assets and display insufficientFundsForProtocolFee copy
+const getHasEnoughBalanceForTxPlusFees = ({
+  balanceCryptoBaseUnit,
+  amountCryptoPrecision,
+  txFeeCryptoBaseUnit,
+  precision,
+}: {
+  balanceCryptoBaseUnit: string
+  amountCryptoPrecision: string
+  txFeeCryptoBaseUnit: string
+  precision: number
+}) => {
+  const balanceCryptoBaseUnitBn = bnOrZero(balanceCryptoBaseUnit)
+  if (balanceCryptoBaseUnitBn.isZero()) return false
+
+  return bnOrZero(amountCryptoPrecision)
+    .plus(fromBaseUnit(txFeeCryptoBaseUnit, precision ?? 0))
+    .lte(fromBaseUnit(balanceCryptoBaseUnitBn, precision))
+}
+
+const getHasEnoughBalanceForTxPlusFeesPlusSweep = ({
+  balanceCryptoBaseUnit,
+  amountCryptoPrecision,
+  txFeeCryptoBaseUnit,
+  precision,
+  sweepTxFeeCryptoBaseUnit,
+}: {
+  balanceCryptoBaseUnit: string
+  amountCryptoPrecision: string
+  txFeeCryptoBaseUnit: string
+  precision: number
+  sweepTxFeeCryptoBaseUnit: string
+}) => {
+  const balanceCryptoBaseUnitBn = bnOrZero(balanceCryptoBaseUnit)
+  if (balanceCryptoBaseUnitBn.isZero()) return false
+
+  return bnOrZero(amountCryptoPrecision)
+    .plus(fromBaseUnit(txFeeCryptoBaseUnit, precision ?? 0))
+    .plus(fromBaseUnit(sweepTxFeeCryptoBaseUnit, precision ?? 0))
+    .lte(fromBaseUnit(balanceCryptoBaseUnitBn, precision ?? 0))
```

Comment:
the `?? 0` might not be required if the type system is working, since precision isnt optional

Reply:
noted that not introduced in this pr, i see its pre-existing

Reply:
e8bb5bdf39